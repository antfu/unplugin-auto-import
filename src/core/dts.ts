import { ImportsFlatMap } from '../types'

export function generateDeclration(imports: ImportsFlatMap, resolvedImports: ImportsFlatMap = {}) {
  const body = [
    ...Object.entries(imports),
    ...Object.entries(resolvedImports),
  ]
    .map(([name, info]) => `  const ${name}: typeof import('${info.module}')['${info.from || name}']`)
    .join('\n')
  return `// Generated by 'unplugin-auto-import'\ndeclare global {\n${body}\n}\nexport {}\n`
}
