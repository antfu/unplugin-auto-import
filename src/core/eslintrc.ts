import type { ESLintGlobalsPropValue, ESLintrc, ImportsFlatMap } from '../types'

interface ESLintConfigs {
  globals: Record<string, ESLintGlobalsPropValue>
}

// const __comment__ = 'Generated by `unplugin-auto-import`'

export function generateESLintConfigs(
  imports: ImportsFlatMap,
  resolvedImports: ImportsFlatMap = {},
  eslintrc: ESLintrc,
) {
  const eslintConfigs: ESLintConfigs = { globals: {} };

  [
    ...Object.entries(imports),
    ...Object.entries(resolvedImports),
  ]
    .sort((a, b) => a[0].localeCompare(b[0]))
    .forEach(([name]) => { eslintConfigs.globals[name] = eslintrc.globalsPropValue || true })
  const jsonBody = JSON.stringify(eslintConfigs, null, 2)
  return jsonBody
}
