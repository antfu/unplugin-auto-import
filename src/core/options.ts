import { resolve } from 'path'
import { toArray } from '@antfu/utils'
import { createFilter } from '@rollup/pluginutils'
import { isPackageExists } from 'local-pkg'
import { presets } from '../presets'
import { ImportsFlatMap, Options, ResolvedOptions, ResolvedResult } from '../types'

export function resolveOptions(options: Options = {}): ResolvedOptions {
  const imports = flattenImportsMap(options.imports, options.presetOverriding)

  const {
    dts = isPackageExists('typescript'),
  } = options

  const resolved: ResolvedOptions = {
    sourceMap: false,
    resolvedImports: {},
    presetOverriding: false,
    ignore: [],
    ...options,
    dts: dts === false
      ? false
      : dts === true
        ? resolve('auto-imports.d.ts')
        : resolve(dts),
    imports,
    resolvers: toArray(options.resolvers),
    idFilter: createFilter(
      options.include || [/\.[jt]sx?$/, /\.vue$/, /\.vue\?vue/, /\.svelte$/],
      options.exclude || [/[\\/]node_modules[\\/]/, /[\\/]\.git[\\/]/],
    ),
  }

  return resolved
}

export function flattenImportsMap(map: Options['imports'], overriding = false): ImportsFlatMap {
  const flat: ImportsFlatMap = {}
  toArray(map).forEach((definition) => {
    if (typeof definition === 'string') {
      if (!presets[definition])
        throw new Error(`[auto-import] preset ${definition} not found`)
      const preset = presets[definition]
      definition = typeof preset === 'function' ? preset() : preset
    }

    for (const mod of Object.keys(definition)) {
      for (const id of definition[mod]) {
        const meta: ResolvedResult = {
          path: mod,
        }
        let name: string
        if (Array.isArray(id)) {
          name = id[1]
          meta.importName = id[0]
        }
        else {
          name = id
          meta.importName = id
        }

        if (flat[name] && !overriding)
          throw new Error(`[auto-import] identifier ${name} already defined with ${flat[name].path}`)

        flat[name] = meta
      }
    }
  })

  return flat
}
