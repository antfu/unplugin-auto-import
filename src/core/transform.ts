import MagicString from 'magic-string'
import { ImportInfo, TransformOptions, Resolver } from '../types'

const excludeRE = [
  // imported from other module
  /\bimport\s*([\w_$]*?),?\s*\{([\s\S]*?)\}\s*from\b/g,
  // defined as function
  /\bfunction\s*([\s\S]+?)\s*\(/g,
  // defined as local variable
  /\b(?:const|let|var)\s*([\w\d_$]+?)\b/g,
]

const matchRE = /\b(\w+)\b/g
const importAsRE = /^.*\sas\s+/
const multilineCommentsRE = /\/\*(.|[\r\n])*?\*\//gm
const singlelineCommentsRE = /\/\/.*/g

function stripeComments(code: string) {
  return code
    .replace(multilineCommentsRE, '')
    .replace(singlelineCommentsRE, '')
}

export function transform(
  code: string,
  id: string,
  {
    imports,
    sourceMap,
    resolvers,
    resolvedImports = {},
  }: TransformOptions,
) {
  const noComments = stripeComments(code)
  const identifiers = new Set(Array.from(noComments.matchAll(matchRE)).map(i => i[1]))

  // nothing matched, skip
  if (!identifiers.size)
    return null

  // remove those already defined
  for (const regex of excludeRE) {
    Array.from(noComments.matchAll(regex))
      .flatMap(i => [...(i[1]?.split(',') || []), ...(i[2]?.split(',') || [])])
      .map(i => i.replace(importAsRE, '').trim())
      .forEach(i => identifiers.delete(i))
  }

  // nothing matched, skip
  if (!identifiers.size)
    return null

  const modules: Record<string, ImportInfo[]> = {}

  // group by module name
  Array.from(identifiers).forEach((name) => {
    let info: ImportInfo = resolvedImports[name] || imports[name]

    if (!info && resolvers?.length) {
      const resolved = firstNonNullResult(resolvers, name)
      if (resolved) {
        if (typeof resolved === 'string') {
          info = {
            module: resolved,
            name,
            from: 'default',
          }
        }
        else {
          info = resolved
        }
        resolvedImports[name] = info
      }
    }

    if (!info)
      return

    if (!modules[info.module])
      modules[info.module] = []
    modules[info.module].push(info)
  })

  if (!Object.keys(modules).length)
    return

  // stringify import
  const importStatements = Object.entries(modules)
    .map(([moduleName, names]) => {
      const imports = names
        .map(({ name, from }) => from ? `${from} as ${name}` : name)
        .join(', ')
      return `import { ${imports} } from '${moduleName}';`
    })
    .join('')

  const s = new MagicString(code)
  s.prependLeft(0, importStatements)

  return {
    code: s.toString(),
    map: sourceMap ? s.generateMap() : null,
  }
}

function firstNonNullResult(array: Resolver[], name: string) {
  for (let i = 0; i < array.length; i++) {
    const res = array[i](name)
    if (res)
      return res
  }
}
