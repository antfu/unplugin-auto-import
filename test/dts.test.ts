import { generateDeclration } from '../src/core/dts'
import { resolveOptions } from '../src/core/options'

it('dts', () => {
  const options = resolveOptions({
    imports: [
      'vue-demi',
      'react',
      'svelte',
      'svelte/animate',
      'svelte/easing',
      'svelte/motion',
      'svelte/store',
      'svelte/transition',
      {
        custom: [
          'customNamed',
          ['default', 'customDefault'],
        ],
      },
    ],
  })

  expect(generateDeclration(options.imports)).toMatchSnapshot()
})
