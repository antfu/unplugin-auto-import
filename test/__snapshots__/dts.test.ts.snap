// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`dts 1`] = `
"// Generated by 'unplugin-auto-import'
declare global {
  const onActivated: typeof import('vue-demi')['onActivated']
  const onBeforeMount: typeof import('vue-demi')['onBeforeMount']
  const onBeforeUnmount: typeof import('vue-demi')['onBeforeUnmount']
  const onBeforeUpdate: typeof import('vue-demi')['onBeforeUpdate']
  const onMounted: typeof import('vue-demi')['onMounted']
  const onUnmounted: typeof import('vue-demi')['onUnmounted']
  const onUpdated: typeof import('vue-demi')['onUpdated']
  const computed: typeof import('vue-demi')['computed']
  const customRef: typeof import('vue-demi')['customRef']
  const isReadonly: typeof import('vue-demi')['isReadonly']
  const isRef: typeof import('vue-demi')['isRef']
  const markRaw: typeof import('vue-demi')['markRaw']
  const reactive: typeof import('vue-demi')['reactive']
  const readonly: typeof import('vue-demi')['readonly']
  const ref: typeof import('vue-demi')['ref']
  const shallowReactive: typeof import('vue-demi')['shallowReactive']
  const shallowReadonly: typeof import('vue-demi')['shallowReadonly']
  const shallowRef: typeof import('vue-demi')['shallowRef']
  const toRaw: typeof import('vue-demi')['toRaw']
  const toRef: typeof import('vue-demi')['toRef']
  const toRefs: typeof import('vue-demi')['toRefs']
  const unref: typeof import('vue-demi')['unref']
  const watch: typeof import('vue-demi')['watch']
  const watchEffect: typeof import('vue-demi')['watchEffect']
  const defineComponent: typeof import('vue-demi')['defineComponent']
  const getCurrentInstance: typeof import('vue-demi')['getCurrentInstance']
  const h: typeof import('vue-demi')['h']
  const inject: typeof import('vue-demi')['inject']
  const nextTick: typeof import('vue-demi')['nextTick']
  const provide: typeof import('vue-demi')['provide']
  const useCssModule: typeof import('vue-demi')['useCssModule']
  const createApp: typeof import('vue-demi')['createApp']
  const useState: typeof import('react')['useState']
  const useCallback: typeof import('react')['useCallback']
  const useMemo: typeof import('react')['useMemo']
  const useEffect: typeof import('react')['useEffect']
  const useRef: typeof import('react')['useRef']
  const useContext: typeof import('react')['useContext']
  const useReducer: typeof import('react')['useReducer']
  const onMount: typeof import('svelte')['onMount']
  const beforeUpdate: typeof import('svelte')['beforeUpdate']
  const afterUpdate: typeof import('svelte')['afterUpdate']
  const onDestroy: typeof import('svelte')['onDestroy']
  const tick: typeof import('svelte')['tick']
  const setContext: typeof import('svelte')['setContext']
  const getContext: typeof import('svelte')['getContext']
  const hasContext: typeof import('svelte')['hasContext']
  const getAllContexts: typeof import('svelte')['getAllContexts']
  const createEventDispatcher: typeof import('svelte')['createEventDispatcher']
  const flip: typeof import('svelte/animate')['flip']
  const linear: typeof import('svelte/easing')['linear']
  const backIn: typeof import('svelte/easing')['backIn']
  const backOut: typeof import('svelte/easing')['backOut']
  const backInOut: typeof import('svelte/easing')['backInOut']
  const bounceIn: typeof import('svelte/easing')['bounceIn']
  const bounceOut: typeof import('svelte/easing')['bounceOut']
  const bounceInOut: typeof import('svelte/easing')['bounceInOut']
  const circIn: typeof import('svelte/easing')['circIn']
  const circOut: typeof import('svelte/easing')['circOut']
  const circInOut: typeof import('svelte/easing')['circInOut']
  const cubicIn: typeof import('svelte/easing')['cubicIn']
  const cubicOut: typeof import('svelte/easing')['cubicOut']
  const cubicInOut: typeof import('svelte/easing')['cubicInOut']
  const elasticIn: typeof import('svelte/easing')['elasticIn']
  const elasticOut: typeof import('svelte/easing')['elasticOut']
  const elasticInOut: typeof import('svelte/easing')['elasticInOut']
  const expoIn: typeof import('svelte/easing')['expoIn']
  const expoOut: typeof import('svelte/easing')['expoOut']
  const expoInOut: typeof import('svelte/easing')['expoInOut']
  const quadIn: typeof import('svelte/easing')['quadIn']
  const quadOut: typeof import('svelte/easing')['quadOut']
  const quadInOut: typeof import('svelte/easing')['quadInOut']
  const quartIn: typeof import('svelte/easing')['quartIn']
  const quartOut: typeof import('svelte/easing')['quartOut']
  const quartInOut: typeof import('svelte/easing')['quartInOut']
  const quintIn: typeof import('svelte/easing')['quintIn']
  const quintOut: typeof import('svelte/easing')['quintOut']
  const quintInOut: typeof import('svelte/easing')['quintInOut']
  const sineIn: typeof import('svelte/easing')['sineIn']
  const sineOut: typeof import('svelte/easing')['sineOut']
  const sineInOut: typeof import('svelte/easing')['sineInOut']
  const tweened: typeof import('svelte/motion')['tweened']
  const spring: typeof import('svelte/motion')['spring']
  const writable: typeof import('svelte/store')['writable']
  const readable: typeof import('svelte/store')['readable']
  const derived: typeof import('svelte/store')['derived']
  const get: typeof import('svelte/store')['get']
  const fade: typeof import('svelte/transition')['fade']
  const blur: typeof import('svelte/transition')['blur']
  const fly: typeof import('svelte/transition')['fly']
  const slide: typeof import('svelte/transition')['slide']
  const scale: typeof import('svelte/transition')['scale']
  const draw: typeof import('svelte/transition')['draw']
  const crossfade: typeof import('svelte/transition')['crossfade']
  const customNamed: typeof import('custom')['customNamed']
  const customDefault: typeof import('custom')['default']
}
export {}
"
`;
